File: src/broker_relay.py
"""Per-jail Claude OAuth broker relay — a raw byte proxy with one trick.

One relay process runs per jail, spawned and supervised by
``src/cli/loopholes_runtime._relay_ensure``.  It listens on
``claude-oauth-broker.sock`` inside the jail's host-services sockets
dir (visible in-jail through the existing directory mount — no
socket-file bind mount) and dials the real broker socket **per
connection**, so a restarted broker (new socket inode) is picked up on
the very next request.  A socket-file bind mount pins the inode — the
"one jail 502s after ``yolo broker restart``" bug — and on macOS
Podman Machine can't bind-mount a socket file at all (EOPNOTSUPP).

It is a standalone process, not a thread in ``yolo run``: conmon keeps
the container alive independently of any single yolo invocation, so
the relay must survive the death of whichever process spawned it.

The one protocol-aware trick: the loophole protocol is exactly one
4-byte-BE length-prefixed UTF-8 JSON request per connection, sent
client-first (``src/host_service.py``).  The relay reads that first
message, stamps ``request["jail_id"]`` with the jail's container name
(host-side injection — trustworthy, unlike an in-jail self-report;
docs/design/loophole-protocol.md: daemons must not trust a client-supplied
value), re-frames it, and then degrades to a dumb bidirectional pipe.
Attribution is best-effort: an unparseable, oversized, or slow first
message is forwarded verbatim and the connection keeps working — the
relay never kills traffic over logging metadata.

Failure semantics the jail-side terminator relies on
(``src/oauth_broker_jail.py``): relay socket missing/refused = relay
layer; relay accepts but ends the connection with zero response frames
= broker layer (the per-connection dial failed).  On dial failure the
relay drains the client's pending request before closing so the client
sees a CLEAN EOF — closing with unread bytes queued would surface as
ECONNRESET (Linux AF_UNIX discards the queue), which the terminator
cannot attribute to a layer.

Not a ``host_service`` handler — this is a byte proxy *in front of*
the frame protocol, not a participant in it.  Logs go to stderr; the
spawner redirects them under ``GLOBAL_STORAGE/logs/``.  Payloads carry
OAuth tokens and are never logged.  SIGTERM unlinks the relay socket
and exits, so a stopped relay reads as "socket absent", not "socket
dead".
"""

from __future__ import annotations

import argparse
import json
import logging
import os
import signal
import socket
import struct
import sys
import threading
import time
from pathlib import Path

log = logging.getLogger("broker_relay")

# First-message bounds for the attribution read.  Requests on this
# protocol are small JSON objects (see host_service._read_request); the
# cap exists so a garbage length prefix can't make us buffer gigabytes,
# and the timeout so a silent client can't park the stamp path forever.
# Blowing either bound downgrades to verbatim forwarding, never failure.
FIRST_MSG_MAX = 4 * 1024 * 1024
FIRST_MSG_TIMEOUT = 5.0


def _pipe(src: socket.socket, dst: socket.socket) -> None:
    """Copy bytes src→dst until EOF or error, then shut down and close.

    Each direction runs one of these; both close BOTH sockets so fds
    never outlive the connection (shutdown() alone doesn't release the
    fd).  The sibling thread's double-close is swallowed.
    """
    try:
        while True:
            data = src.recv(65536)
            if not data:
                break
            dst.sendall(data)
    except OSError:
        pass
    finally:
        for s, how in [(src, socket.SHUT_RD), (dst, socket.SHUT_WR)]:
            try:
                s.shutdown(how)
            except OSError:
                pass
        for s in (src, dst):
            try:
                s.close()
            except OSError:
                pass


def _read_first_message(client: socket.socket) -> "tuple[bytes | None, bytes]":
    """Try to read the connection's single framed request.

    Returns ``(body, raw)`` where ``raw`` is every byte consumed so far.
    ``body`` is the frame payload iff a complete frame arrived within
    FIRST_MSG_TIMEOUT / FIRST_MSG_MAX; on EOF, timeout, or an oversized
    length prefix it is None and the caller forwards ``raw`` verbatim.
    """
    raw = bytearray()
    client.settimeout(FIRST_MSG_TIMEOUT)
    try:
        while len(raw) < 4:
            chunk = client.recv(4 - len(raw))
            if not chunk:
                return None, bytes(raw)
            raw += chunk
        (length,) = struct.unpack(">I", raw[:4])
        if length > FIRST_MSG_MAX:
            return None, bytes(raw)
        while len(raw) < 4 + length:
            chunk = client.recv(4 + length - len(raw))
            if not chunk:
                return None, bytes(raw)
            raw += chunk
        return bytes(raw[4:]), bytes(raw)
    except OSError:
        # Timeout or client error mid-message — hand back what we have.
        return None, bytes(raw)
    finally:
        client.settimeout(None)


def _stamp_jail_id(body: bytes, jail_id: str) -> "bytes | None":
    """Re-frame ``body`` with ``jail_id`` stamped, or None if it isn't a
    JSON object (caller then forwards the original bytes verbatim).

    The stamp OVERRIDES any client-supplied jail_id — attribution must
    come from the host side, not from the jail naming itself.
    """
    try:
        request = json.loads(body.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError):
        return None
    if not isinstance(request, dict):
        return None
    request["jail_id"] = jail_id
    new_body = json.dumps(request).encode("utf-8")
    return struct.pack(">I", len(new_body)) + new_body


def _handle(client: socket.socket, broker_path: str, jail_id: str) -> None:
    """Serve one client connection: dial the broker, stamp, pipe."""
    try:
        upstream = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        upstream.connect(broker_path)
    except OSError as e:
        # Broker layer: end the connection with a clean EOF and zero
        # response frames — the terminator reads EOF-before-exit-frame
        # as "broker unreachable through the relay".  Closing while the
        # client's request sits unread in our receive queue would
        # DISCARD those bytes and raise ECONNRESET at the peer (Linux
        # AF_UNIX) instead of EOF, so the terminator could never
        # attribute the failure to a layer.  Order matters: shut down
        # our write side first (the client's recv sees EOF and it
        # closes), then drain whatever it already sent, then close.
        log.warning("dial %s failed: %s", broker_path, e)
        try:
            client.shutdown(socket.SHUT_WR)
        except OSError:
            pass
        try:
            # Bounded drain — a wedged client can't park this thread.
            deadline = time.monotonic() + FIRST_MSG_TIMEOUT
            while time.monotonic() < deadline:
                client.settimeout(max(0.05, deadline - time.monotonic()))
                if not client.recv(65536):
                    break
        except OSError:
            pass
        try:
            client.close()
        except OSError:
            pass
        return

    body, raw = _read_first_message(client)
    framed = _stamp_jail_id(body, jail_id) if body is not None else None
    if framed is None:
        if raw:
            # Payloads may carry tokens — log the length, never the bytes.
            log.warning(
                "first message not a framed JSON object (%d bytes) — "
                "forwarding unstamped",
                len(raw),
            )
        framed = raw
    try:
        if framed:
            upstream.sendall(framed)
    except OSError:
        for s in (client, upstream):
            try:
                s.close()
            except OSError:
                pass
        return

    threading.Thread(target=_pipe, args=(client, upstream), daemon=True).start()
    _pipe(upstream, client)


def serve(socket_path: Path, broker_path: Path, jail_id: str) -> int:
    """Accept-loop until SIGTERM/SIGINT; one handler thread per client."""
    socket_path.parent.mkdir(parents=True, exist_ok=True)
    # A stale file at our path (crashed predecessor) would EADDRINUSE.
    socket_path.unlink(missing_ok=True)
    srv = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    bound_id: "tuple[int, int] | None" = None
    try:
        srv.bind(str(socket_path))
        # Remember which FILE we bound, so exit cleanup can tell "our
        # socket" from a successor's that re-bound the same path.
        try:
            st = os.stat(socket_path)
            bound_id = (st.st_dev, st.st_ino)
        except OSError:
            bound_id = None
        srv.listen(32)
        log.info("relaying %s -> %s (jail=%s)", socket_path, broker_path, jail_id)

        stop = threading.Event()

        def _graceful(signo, _frame):
            log.info("signal %d — shutting down", signo)
            stop.set()
            try:
                srv.shutdown(socket.SHUT_RDWR)
            except OSError:
                # BSD/macOS: shutdown() on a LISTENING socket raises
                # ENOTCONN, and PEP 475 transparently retries the
                # interrupted accept() — the loop would never observe
                # ``stop``.  close() below is what actually breaks
                # accept() out there (EBADF); on Linux it's a no-op
                # after the successful shutdown.
                pass
            try:
                srv.close()
            except OSError:
                pass

        # Signal handlers can only be installed from the main thread;
        # they always are when we run as the spawned daemon process.
        if threading.current_thread() is threading.main_thread():
            signal.signal(signal.SIGTERM, _graceful)
            signal.signal(signal.SIGINT, _graceful)

        while not stop.is_set():
            try:
                conn, _ = srv.accept()
            except OSError:
                break
            threading.Thread(
                target=_handle,
                args=(conn, str(broker_path), jail_id),
                daemon=True,
            ).start()
    finally:
        try:
            srv.close()
        except OSError:
            pass
        # The socket file must not outlive the process: a leftover file
        # reads as "relay dead" (ECONNREFUSED) instead of "relay absent"
        # to doctor and the terminator.  But only unlink the file WE
        # bound: if this relay leaked (e.g. its /tmp pidfile was aged
        # away and a successor healed over it, re-binding the same
        # path), the path now names the successor's LIVE socket and
        # removing it would 502 the jail until the next heal.
        try:
            st = os.stat(socket_path)
            if bound_id is not None and (st.st_dev, st.st_ino) == bound_id:
                socket_path.unlink()
        except OSError:
            pass
    return 0


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Per-jail relay: pipes a jail's broker socket to the "
        "host-wide singleton, dialing it per connection."
    )
    parser.add_argument(
        "--socket",
        required=True,
        type=Path,
        help="relay listen socket (inside the jail's host-services dir)",
    )
    parser.add_argument(
        "--broker",
        required=True,
        type=Path,
        help="real broker socket, dialed per connection",
    )
    parser.add_argument(
        "--jail",
        required=True,
        help="container name stamped as jail_id on each request",
    )
    args = parser.parse_args()
    logging.basicConfig(
        stream=sys.stderr,
        level=logging.INFO,
        format="%(asctime)s %(levelname)s %(name)s: %(message)s",
    )
    return serve(args.socket, args.broker, args.jail)


if __name__ == "__main__":
    sys.exit(main())

File: internal/brokerrelay/brokerrelay.go
// Package brokerrelay is the per-jail Claude OAuth broker relay — a raw byte
// proxy with one protocol-aware trick — ported from src/broker_relay.py.
//
// One relay runs per jail, spawned by loopholes_runtime._relay_ensure. It
// listens on a HOST-ONLY Unix socket and dials the real broker socket PER
// CONNECTION, so a restarted broker (new socket inode) is picked up on the very
// next request (the "one jail 502s after `yolo broker restart`" bug the
// per-connection dial fixes).
//
// # How the jail reaches it
//
// NOT through that socket. The jail's hop is loopback-TLS (internal/svcendpoint):
// a front terminates a pinned-certificate, token-authenticated connection and
// SPLICES it into the Unix socket below. Two consequences worth stating:
//
//   - the Unix socket is host-only on purpose. It lives beside this relay's pid
//     and lock files in /tmp, NOT in the jail's :rw-mounted host-services dir —
//     leaving it there would keep the retired transport reachable from inside the
//     jail, and would let the jail unlink the relay's own socket. That directory
//     now holds endpoint files and nothing else;
//   - the relay core below is UNTOUCHED by the transport. It stays typed on
//     *net.UnixConn, which is why the front splices instead of swapping the
//     listener: pipe()'s teardown is frozen Python parity and a tls.Conn has
//     CloseWrite but no CloseRead.
//
// The one protocol-aware trick: the loophole protocol is exactly one 4-byte-BE
// length-prefixed UTF-8 JSON request per connection, client-first. The relay
// reads that first message, stamps request["jail_id"] with the jail's
// container name (host-side injection — trustworthy, unlike an in-jail
// self-report), re-frames it, then degrades to a dumb bidirectional pipe.
// Attribution is best-effort: an unparseable / oversized / slow first message
// is forwarded verbatim and the connection keeps working.
//
// Failure semantics the jail-side terminator relies on: the endpoint file
// missing, or the address in it refusing the dial, = relay layer; the relay
// accepts but ends the connection with zero response frames = broker layer. On
// dial failure the relay drains the
// client's pending request before closing so the client sees a CLEAN EOF —
// closing with unread bytes queued surfaces as ECONNRESET (Linux AF_UNIX
// discards the rx queue), which the terminator cannot attribute to a layer.
//
// SIGTERM unlinks the relay socket ONLY if its dev/ino still match what we
// bound (so a successor that healed over the same path is never disturbed),
// then exits — a stopped relay reads as "socket absent", not "socket dead".
package brokerrelay

import (
	"encoding/binary"
	"io"
	"log"
	"net"
	"os"
	"sync"
	"syscall"
	"time"

	"github.com/mschulkind-oss/yolo-jail/internal/jsonx"
	"github.com/mschulkind-oss/yolo-jail/internal/svcendpoint"
)

// First-message bounds for the attribution read. Requests are small JSON
// objects; the cap stops a garbage length prefix from buffering gigabytes, the
// timeout stops a silent client from parking the stamp path forever. Blowing
// either bound downgrades to verbatim forwarding, never failure.
const (
	firstMsgMax     = 4 * 1024 * 1024
	firstMsgTimeout = 5 * time.Second
)

// Logger is where the relay writes its (payload-free) diagnostics. Payloads
// carry OAuth tokens and are NEVER logged.
var Logger = log.New(os.Stderr, "", log.LstdFlags)

// pipe copies src->dst until EOF or error, then shuts down and closes BOTH
// sockets so fds never outlive the connection (shutdown alone doesn't release
// the fd). The sibling goroutine's double-close is swallowed.
func pipe(src, dst *net.UnixConn, wg *sync.WaitGroup) {
	if wg != nil {
		defer wg.Done()
	}
	buf := make([]byte, 65536)
	for {
		n, err := src.Read(buf)
		if n > 0 {
			if _, werr := dst.Write(buf[:n]); werr != nil {
				break
			}
		}
		if err != nil {
			break
		}
	}
	// Match Python's finally: shut down src read side + dst write side, then
	// close both. Errors are ignored (the sibling may have closed already).
	_ = src.CloseRead()
	_ = dst.CloseWrite()
	_ = src.Close()
	_ = dst.Close()
}

// readFirstMessage tries to read the connection's single framed request.
// Returns (body, raw): raw is EVERY byte consumed so far; body is the frame
// payload iff a complete frame arrived within firstMsgTimeout / firstMsgMax,
// else nil and the caller forwards raw verbatim. Faithful to Python's
// _read_first_message, which accumulates raw byte-by-byte via recv, so on EOF
// or timeout at any offset raw holds exactly what was received (io.ReadFull's
// n-on-error is captured here for that reason).
func readFirstMessage(client *net.UnixConn) (body, raw []byte) {
	_ = client.SetReadDeadline(time.Now().Add(firstMsgTimeout))
	defer client.SetReadDeadline(time.Time{})

	header := make([]byte, 4)
	n, err := io.ReadFull(client, header)
	raw = append(raw, header[:n]...)
	if err != nil {
		return nil, raw // short/clean-EOF header — forward what we got
	}
	length := binary.BigEndian.Uint32(header)
	if length > firstMsgMax {
		return nil, raw // oversized length prefix — forward the header verbatim
	}
	payload := make([]byte, length)
	n, err = io.ReadFull(client, payload)
	raw = append(raw, payload[:n]...)
	if err != nil {
		return nil, raw // partial body (timeout/EOF mid-message)
	}
	return payload, raw
}

// stampJailID re-frames body with jailID stamped, or returns (nil,false) if it
// isn't a JSON object (caller then forwards the original bytes verbatim). The
// stamp OVERRIDES any client-supplied jail_id — attribution must come from the
// host side. Re-serialization uses jsonx (Python json.dumps default separators
// + insertion order) so the re-framed request matches what Python would emit.
func stampJailID(body []byte, jailID string) ([]byte, bool) {
	decoded, err := jsonx.Decode(body)
	if err != nil {
		return nil, false
	}
	m, ok := decoded.(*jsonx.OrderedMap)
	if !ok {
		return nil, false
	}
	m.Set("jail_id", jailID) // insert-or-update, preserving position
	newBody, err := jsonx.DumpsCompact(m)
	if err != nil {
		return nil, false
	}
	nb := []byte(newBody)
	framed := make([]byte, 4+len(nb))
	binary.BigEndian.PutUint32(framed[:4], uint32(len(nb)))
	copy(framed[4:], nb)
	return framed, true
}

// handle serves one client connection: dial the broker, stamp, pipe.
func handle(client *net.UnixConn, brokerPath, jailID string) {
	upstream, err := net.DialUnix("unix", nil, &net.UnixAddr{Name: brokerPath, Net: "unix"})
	if err != nil {
		// Broker layer: clean EOF + zero frames. Shut down our write side
		// (client's recv sees EOF), drain what it already sent, then close —
		// closing with the request unread would DISCARD it and raise
		// ECONNRESET at the peer instead of EOF.
		Logger.Printf("dial %s failed: %v", brokerPath, err)
		_ = client.CloseWrite()
		drainBounded(client, firstMsgTimeout)
		_ = client.Close()
		return
	}

	body, raw := readFirstMessage(client)
	var framed []byte
	if body != nil {
		if stamped, ok := stampJailID(body, jailID); ok {
			framed = stamped
		}
	}
	if framed == nil {
		if len(raw) > 0 {
			// Payloads may carry tokens — log the length, never the bytes.
			Logger.Printf("first message not a framed JSON object (%d bytes) — forwarding unstamped", len(raw))
		}
		framed = raw
	}

	if len(framed) > 0 {
		if _, err := upstream.Write(framed); err != nil {
			_ = client.Close()
			_ = upstream.Close()
			return
		}
	}

	var wg sync.WaitGroup
	wg.Add(1)
	go pipe(client, upstream, &wg)
	pipe(upstream, client, nil)
	wg.Wait()
}

// drainBounded reads and discards from conn until EOF or the deadline — a
// wedged client can't park the goroutine.
func drainBounded(conn *net.UnixConn, budget time.Duration) {
	deadline := time.Now().Add(budget)
	buf := make([]byte, 65536)
	for {
		now := time.Now()
		if !now.Before(deadline) {
			return
		}
		_ = conn.SetReadDeadline(deadline)
		n, err := conn.Read(buf)
		if n == 0 || err != nil {
			return
		}
	}
}

// Config is one relay's parameters. EndpointPath and AdvertiseHost are the ONLY
// additions the loopback-TLS migration makes to this package; everything else is
// the frozen relay core.
type Config struct {
	// SocketPath is the relay's own Unix socket. HOST-ONLY — see the package doc.
	SocketPath string
	// BrokerPath is the host-wide broker singleton, dialed per connection.
	BrokerPath string
	// JailID is stamped host-side onto every request.
	JailID string
	// EndpointPath, when non-empty, is where the loopback-TLS front publishes its
	// address, its certificate and this jail's bearer token.
	//
	// THAT FILE IS A CREDENTIAL: 0600, in the per-jail host-services dir, never
	// shared. Empty means no front at all — the relay then serves the Unix socket
	// only, which is what this package's own tests use and what a host-side caller
	// on the same machine can still reach.
	EndpointPath string
	// AdvertiseHost overrides the host name the front publishes for the jail to
	// dial. Empty resolves through svcendpoint.AdvertiseHost(), i.e. the
	// YOLO_SVC_ADVERTISE_HOST this process was spawned with, else the container
	// runtime's gateway name.
	AdvertiseHost string
}

// Serve runs the accept loop until stop is closed; one goroutine per client.
// dev/ino, and on shutdown unlink the socket ONLY if it's still the file we
// bound. Returns nil on clean shutdown.
func Serve(cfg Config, stop <-chan struct{}) error {
	// Destructured up front so the relay core below is TEXTUALLY unchanged by the
	// transport work — a diff that touches pipe()'s teardown or the stamp path is
	// then visibly not this change.
	socketPath, brokerPath, jailID := cfg.SocketPath, cfg.BrokerPath, cfg.JailID
	if err := os.MkdirAll(dir(socketPath), 0o755); err != nil {
		return err
	}
	// A stale file at our path (crashed predecessor) would EADDRINUSE.
	_ = os.Remove(socketPath)

	ln, err := net.ListenUnix("unix", &net.UnixAddr{Name: socketPath, Net: "unix"})
	if err != nil {
		return err
	}
	// The jail's half of the hop, started only once the socket it splices into
	// exists. FIRE AND FORGET, deliberately: a front that cannot bind or cannot
	// publish logs and dies while Serve keeps serving the Unix socket. Failing
	// Serve instead would take down a relay a same-host caller can still use, and
	// would report a publication problem as "the relay is dead" — a worse diagnosis
	// of the same fact. The host-side health gate (relayHealthy) sees the absent
	// endpoint and respawns, and `yolo check` names it.
	//
	// NoPreamble: THE ONE DELIBERATE OPT-OUT IN THE TREE, and it is on the
	// credential path. The relay is not a dumb pipe — handle -> readFirstMessage
	// reads the FIRST length-prefixed frame off this connection and stampJailID
	// rewrites it. A connection preamble (svcendpoint/preamble.go) is framed
	// identically, so with one in front the relay would stamp THE PREAMBLE, write
	// that to the broker, and leave the terminator's real request sitting unread
	// behind it: every jail's Claude OAuth refresh fails. The design doc called
	// this change "additive"; it is not, and this line is the whole correction
	// (docs/design/broker-as-a-pack.md §12, retraction).
	//
	// REMOVAL TRIGGER: the broker conversion, which deletes readFirstMessage and
	// stampJailID outright — the preamble is what replaces them. When this
	// function goes, so does the opt-out; until then nothing else may set it.
	if cfg.EndpointPath != "" {
		go func() {
			if err := svcendpoint.ServeFrontWithOptions(cfg.EndpointPath, cfg.AdvertiseHost, socketPath, stop,
				svcendpoint.FrontOptions{NoPreamble: true}); err != nil {
				Logger.Printf("front: %s not published (%v) — this jail cannot reach the relay",
					cfg.EndpointPath, err)
			}
		}()
	}
	// Go's UnixListener unlinks the socket on Close by default; disable that so
	// WE control unlink (only-if-ours), matching Python's dev/ino guard.
	ln.SetUnlinkOnClose(false)

	boundDev, boundIno, haveBound := statDevIno(socketPath)

	Logger.Printf("relaying %s -> %s (jail=%s)", socketPath, brokerPath, jailID)

	// Closing the listener breaks Accept out of its blocking call.
	go func() {
		<-stop
		_ = ln.Close()
	}()

	for {
		conn, err := ln.AcceptUnix()
		if err != nil {
			// stop closed => listener closed => break cleanly.
			break
		}
		go handle(conn, brokerPath, jailID)
	}

	// Cleanup: unlink the socket file only if it's still the one we bound.
	if haveBound {
		if dev, ino, ok := statDevIno(socketPath); ok && dev == boundDev && ino == boundIno {
			_ = os.Remove(socketPath)
		}
	}
	return nil
}

// statDevIno returns the (dev, ino) of path, or ok=false on error.
func statDevIno(path string) (dev uint64, ino uint64, ok bool) {
	var st syscall.Stat_t
	if err := syscall.Stat(path, &st); err != nil {
		return 0, 0, false
	}
	return uint64(st.Dev), st.Ino, true
}

func dir(path string) string {
	for i := len(path) - 1; i >= 0; i-- {
		if path[i] == '/' {
			if i == 0 {
				return "/"
			}
			return path[:i]
		}
	}
	return "."
}
